src/watchdog.js

```diff
@@ -0,0 +1,401 @@
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Saves the editor data, so it can be restored after the crash even if the data cannot be fetched at
+	 * the moment of a crash.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
+			this._throttledSave.cancel();
+
+			return;
+		}
+
+		try {
+			this._data = this._editor.getData();
+			this._lastDocumentVersion = version;
+		} catch ( err ) {
+			console.error(
+				err,
+				'An error happened during restoring editor data. ' +
+				'Editor will be restored from the previously saved data.'
+			);
+		}
+	}
+
+	/**
+	 * Checks if the event error comes from the editor that is handled by the watchdog (by checking the error context) and
+	 * restarts the editor. It handles {@link module:utils/ckeditorerror~CKEditorError CKEditorError errors} only.
+	 *
+	 * @private
+	 * @fires error
+	 * @param {Event} event Error event.
+	 */
+	_handleGlobalErrorEvent( event ) {
+		if ( !event.error.is || !event.error.is( 'CKEditorError' ) ) {
+			return;
+		}
+
+		if ( event.error.context === undefined ) {
+			console.error( 'The error is missing its context and Watchdog cannot restart the proper editor.' );
+
+			return;
+		}
+
+		// In some cases the editor should not be restarted - e.g. in case of the editor initialization.
+		// That's why the `null` was introduced as a correct error context which does cause restarting.
+		if ( event.error.context === null ) {
+			return;
+		}
+
+		if ( this._isErrorComingFromThisEditor( event.error ) ) {
+			this.crashes.push( {
+				message: event.error.message,
+				source: event.source,
+				lineno: event.lineno,
+				colno: event.colno
+			} );
+
+			this.fire( 'error' );
+
+			if ( this.crashes.length <= this._crashNumberLimit ) {
+				this.restart();
+			}
+		}
+	}
+
+	/**
+	 * Traverses both structures to find out whether the error context is connected
+	 * with the current editor.
+	 *
+	 * @private
+	 * @param {module:utils/ckeditorerror~CKEditorError} error
+	 */
+	_isErrorComingFromThisEditor( error ) {
+		return areConnectedThroughProperties( this._editor, error.context );
+	}
+
+	/**
+	 * A shortcut method for creating the Watchdog. For the full usage see the
+	 * {@link ~Watchdog Watchdog class description}.
+	 *
+	 * Usage:
+	 *
+	 * 		const watchdog = Watchdog.for( ClassicEditor );
+	 *
+	 * 		watchdog.create( elementOrData, config ).then( editor => {} );
+	 *
+	 * @param {*} Editor The editor class.
+	 */
+	static for( Editor ) {
+		const watchdog = new Watchdog();
+
+		watchdog.setCreator( ( elementOrData, config ) => Editor.create( elementOrData, config ) );
+		watchdog.setDestructor( editor => editor.destroy() );
+
+		return watchdog;
+	}
+
+	/**
+	 * Fired when an error occurs and the watchdog will be restarting the editor.
+	 *
+	 * @event error
+	 */
```

Comment:
Actually not sure if we need this event. It duplicates the `restart` event a little.

Reply:
But at this event, you can add some actions strictly connecting to error handling. E.g.:
- blurring the editing area
- some cleanup

Note that the `restart` event fires even after manually calling the `restart` method.

Reply:
Ok, then, we can keep the `error` event.